core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_MDRange.hpp

```diff
@@ -0,0 +1,1590 @@
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_OPENACC_PARALLEL_REDUCE_MDRANGE_HPP
+#define KOKKOS_OPENACC_PARALLEL_REDUCE_MDRANGE_HPP
+
+#include <OpenACC/Kokkos_OpenACC.hpp>
+#include <OpenACC/Kokkos_OpenACC_FunctorAdapter.hpp>
+#include <OpenACC/Kokkos_OpenACC_MDRangePolicy.hpp>
+#include <Kokkos_Parallel.hpp>
+#include <type_traits>
+
+namespace Kokkos::Experimental::Impl {
+
+using OpenACCIterateLeft  = std::integral_constant<Iterate, Iterate::Left>;
+using OpenACCIterateRight = std::integral_constant<Iterate, Iterate::Right>;
+template <int N>
+using OpenACCMDRangeBegin = decltype(MDRangePolicy<OpenACC, Rank<N>>::m_lower);
+template <int N>
+using OpenACCMDRangeEnd = decltype(MDRangePolicy<OpenACC, Rank<N>>::m_upper);
+
+// primary template: catch-all non-implemented custom reducers
+template <class Functor, class Reducer, class Policy,
+          bool = std::is_arithmetic_v<typename Reducer::value_type>>
+struct OpenACCParallelReduceMDRangeHelper {
+  OpenACCParallelReduceMDRangeHelper(Functor const&, Reducer const&,
+                                     Policy const&) {
+    static_assert(
+        !Kokkos::Impl::always_true<Functor>::value,
+        "Kokkos Error: not implemented reducer type for the OpenACC backend");
+  }
+};
+
+}  // namespace Kokkos::Experimental::Impl
+
+template <class Functor, class ReducerType, class... Traits>
+class Kokkos::Impl::ParallelReduce<Functor, Kokkos::MDRangePolicy<Traits...>,
+                                   ReducerType, Kokkos::Experimental::OpenACC> {
+  using Policy = MDRangePolicy<Traits...>;
+
+  using Analysis =
+      FunctorAnalysis<FunctorPatternInterface::REDUCE, Policy, Functor>;
+
+  using Pointer   = typename Analysis::pointer_type;
+  using ValueType = typename Analysis::value_type;
+
+  Functor m_functor;
+  Policy m_policy;
+  ReducerType m_reducer;
+  Pointer m_result_ptr;
+
+ public:
+  ParallelReduce(Functor const& functor, Policy const& policy,
+                 ReducerType const& reducer)
+      : m_functor(functor),
+        m_policy(policy),
+        m_reducer(reducer),
+        m_result_ptr(reducer.view().data()) {}
+
+  template <class ViewType>
+  ParallelReduce(
+      const Functor& functor, const Policy& policy, const ViewType& result,
+      std::enable_if_t<Kokkos::is_view<ViewType>::value, void*> = nullptr)
+      : m_functor(functor),
+        m_policy(policy),
+        m_reducer(InvalidType()),
+        m_result_ptr(result.data()) {}
+
+  void execute() {
+    static_assert(Policy::rank < 7 && Policy::rank > 1,
+                  "OpenACC Backend MDRangePolicy Error: Unsupported rank...");
```

Comment:
Shouldn't this check be in the `MDRangePolicy` class? Is this restriction specific to the `OpenACC` backend?

Reply:
This is just for double-checking for the OpenACC backend. 
Whenever rank restriction in MDRangePolicy is changed, the OpenACC backend implementation should be changed too. This static_assert can detect unintended mismatch between rank restrictions in MDRangePolicy and the OpenACC backend implementation.